### DISCUSSION TITLE
Out of memory errors when INSERT SELECT

### DISCUSSION THREAD

**Discussion by mfridental** [ClickHouse/ClickHouse#60743] [General]
The whole last week I was struggling with exporting a 15 TB table to S3 always getting out or memory errors. Which leads me to the question, why are out of memory errors even possible when executing INSERT SELECT? ClickHouse could read just as many rows as it fits in half of the memory, write them to the target, free memory, repeat. Why aren't we doing this? No other queries are running at the node. Nothing funny happens in the select query (no joins, etc). Is there any (combination) of settings that would lead to the behavior described above? I am currently using 
```
settings output_format_parquet_compression_method='gzip',
            min_bytes_to_use_mmap_io=0,
            s3_truncate_on_insert=1,
            max_insert_delayed_streams_for_parallel_write=6
```
which is still not helpful.

**Comment by ArctypeZach** [comment]
Hi @mfridental - check out https://clickhouse.com/blog/supercharge-your-clickhouse-data-loads-part2

Specifically, try running the insert with `min_insert_block_size_rows=0` and setting `min_insert_block_size_bytes`, `peak_memory_usage_in_bytes`, and `max_insert_threads` according to the formula mentioned:

> • `max_insert_threads`: choose ~ half of the available CPU cores for insert threads (to leave enough dedicated cores for background merges)
> 
> • `peak_memory_usage_in_bytes`: choose an intended peak memory usage; either all available RAM (if it is an isolated ingest) or half or less (to leave room for other concurrent tasks)
> 
> Then:
> `min_insert_block_size_bytes` = `peak_memory_usage_in_bytes` / (~3 * `max_insert_threads`)
